### PR TITLE
[SPARK-4621][shuffle]Shuffle index can be cached for SortShuffleManager in ExternalShuffle in order to  reduce indexFile's io

### PR DESCRIPTION
**Pull request by lianhuiwang**
In ExternalShuffle shuffle index can cached for SortShuffleManager at first. Next time another reduce's fetching can read index from cache. Because reduce tasks fetch map output from ExternalShuffle at same time, shuffle index cache is effective.
In my test that it has 1000 partitions and 100 map tasks, it spends 3882ms for no cache index and it spends 608ms for cache index. So it is six times faster than index file.
@rxin @JoshRosen @andrewor14

### CODE REVIEW

**Comment by andrewor14** [comment]
@lianhuiwang can you make this pass tests first?

File: network/shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleIndexCache.java [hunk 1]
```diff
@@ -0,0 +1,251 @@
+package org.apache.spark.network.shuffle;
```
Comment (andrewor14) [inline network/shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleIndexCache.java]: need license header

File: network/shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleIndexRecord.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/** The offsets of a single map output file for a reduceId */
+public class ShuffleIndexRecord {
+  public long offset;
+  public long nextOffset;
+
+  public ShuffleIndexRecord(long offset, long nextOffset) {
+    this.offset = offset;
+    this.nextOffset = nextOffset;
+  }
+}
```
Comment (andrewor14) [inline network/shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleIndexRecord.java]: need new line after

File: network/shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleIndexCache.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+      }
+
+      return info.getIndex(reduceId);
+    } else {
+      return this.readIndexFile(executor, shuffleId, mapId, reduceId);
+    }
+  }
+
+  public ShuffleIndexRecord readIndexFile(
+    ExecutorShuffleInfo executor, int shuffleId, int mapId, int reduceId) throws IOException {
```
Comment (andrewor14) [inline network/shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleIndexCache.java]: style:
```
public ShuffleIndexRecord readIndexFile(
    ExecutorShuffleInfo executor,
    int shuffleId,
    int mapId,
    int reduceId) throws IOException {
  ...
}
```

**Comment by andrewor14** [comment]
Also why not do it for the non-external case as well?

**Comment by lianhuiwang** [comment]
@andrewor14 thanks. because now we only use external shuffle. but it does not matter. latter i will add non-external in this PR.

**Comment by andrewor14** [comment]
actually, I would do it in a separate PR. In general it's good to keep these small.

**Comment by lianhuiwang** [comment]
@andrewor14 I have update this PR with non-external shuffle.
Now use two configurations: 
spark.shuffle.index.cacheSize is for non-external, spark.shuffle.service.indexCacheSize is for external. 
I am not sure they can be the same one configuration, because their size is different. So now i use two different configuration.

**Comment by lianhuiwang** [comment]
Now i set cacheSize=0 in ExternalShuffleServiceSuite. because every unit test use the same shuffleId at the same time with multi-threads.

**Pull request closed without merging**